source/fe/fe_p1nc.cc

```diff
@@ -0,0 +1,418 @@
+            {
+              hessians[k][0][0] = 0.0 ;
+              hessians[k][0][1] = 0.0 ;
+              hessians[k][1][0] = 0.0 ;
+              hessians[k][1][1] = 0.0 ;
+              output_data.shape_hessians[k][i] = hessians[k];
+            }
+        }
+    }
+
+  // When this function is called for the graphic out,
+  // MappingRelatedData does not work properly.
+  // In this case, the quadrature points on the real cell is computed in manual sense, using 'mapping' and 'quadrature'.
+  // This is a temporary solution. It needs to be fixed fundamentally.
+  if (n_q_points==0)
+    {
+      for (unsigned int i=0; i<quadrature.size(); ++i)
+        for (unsigned int k=0; k<this->dofs_per_cell; ++k)
+          {
+            Point<2> realquadrature ;
+
+            realquadrature = mapping.transform_unit_to_real_cell(cell, quadrature.point(i)) ;
+            values[k] = a[k]*realquadrature(0) + b[k]*realquadrature(1) + c[k] ;
+            output_data.shape_values[k][i] = values[k];
+          }
+    }
+}
+
+
+
+void
+FE_P1NC::fill_fe_face_values (const Triangulation<2,2>::cell_iterator           &cell,
+                              const unsigned int                                 face_no,
+                              const Quadrature<1>                                             &quadrature,
+                              const Mapping<2,2>                                         &mapping,
+                              const Mapping<2,2>::InternalDataBase              &mapping_internal,
+                              const dealii::internal::FEValues::MappingRelatedData<2,2> &mapping_data,
+                              const InternalDataBase                                              &fe_internal,
+                              dealii::internal::FEValues::FiniteElementRelatedData<2,2> &output_data) const
+
+{
+  const UpdateFlags flags(fe_internal.update_each) ;
+
+  const unsigned int n_q_points = mapping_data.quadrature_points.size();
+
+  std::vector<double> values(flags & update_values ? this->dofs_per_cell : 0);
+  std::vector<Tensor<1,2> > grads(flags & update_gradients ? this->dofs_per_cell : 0);
+
+
+  // linear shape with a half value: phi(x,y) = ax + by + c
+  std::vector<double> a(4), b(4), c(4);
+  get_linear_shape (cell, a, b, c);
+
+
+  // compute basis functions
+  if (flags & (update_values | update_gradients))
+    for (unsigned int i=0; i<n_q_points; ++i)
+      {
+        for (unsigned int k=0; k<this->dofs_per_cell; ++k)
+          {
+            if (flags & update_values)
+              {
+                values[k] = a[k]*mapping_data.quadrature_points[i](0) + b[k]*mapping_data.quadrature_points[i](1) + c[k] ;
+                output_data.shape_values[k][i] = values[k];
+              }
+
+            if (flags & update_gradients)
+              {
+                grads[k][0] = a[k] ;
+                grads[k][1] = b[k] ;
+                output_data.shape_gradients[k][i] = grads[k];
+              }
```

Comment:
you forget to handle the Hessians here, but that would not be necessary if you did as suggested above.